binding/ios/PorcupineManager.swift

```diff
@@ -11,150 +11,234 @@
+        try self.init(modelFilePath: modelFilePath, wakeKeywordConfigurations: [wakeKeywordConfiguration], onDetection: onDetection)
+    }
+
+    deinit {
+        pv_porcupine_delete(porcupine)
+        porcupine = nil
+        if isListening {
+            stopListening()
+        }
     }
 
-    /// Initializes Porcupine engine and audio recording.
+    /// Start listening for configured wake words.
     ///
-    /// - Throws: Throws 'PorcupineManagerError' when Porcupine initialization fails.
-    func start() throws {
-        if isListening { return }
+    /// - Throws: AVAudioSession, AVAudioEngine errors. Additionally PorcupineManagerPermissionError if
+    ///           microphone permission is not granted.
+    public func startListening() throws {
 
-        let status = pv_porcupine_multiple_keywords_init(
-            modelFilePath,
-            Int32(keywordFilePaths.count),
-            keywordFilePaths.map{ UnsafePointer(strdup($0)) },
-            sensitivities,
-            &porcupine)
-         try checkStatus(status)
-
-        var format = AudioStreamBasicDescription(
-            mSampleRate: Float64(pv_sample_rate()),
-            mFormatID: kAudioFormatLinearPCM,
-            mFormatFlags: kLinearPCMFormatFlagIsSignedInteger | kLinearPCMFormatFlagIsPacked,
-            mBytesPerPacket: 2,
-            mFramesPerPacket: 1,
-            mBytesPerFrame: 2,
-            mChannelsPerFrame: 1,
-            mBitsPerChannel: 16,
-            mReserved: 0)
-        let userData = UnsafeMutableRawPointer(Unmanaged.passUnretained(self).toOpaque())
-        AudioQueueNewInput(&format, audioCallback, userData, nil, nil, 0, &queue)
-
-        guard let queue = queue else { return }
-        let bufferSize = UInt32(pv_porcupine_frame_length()) * 2
-        for _ in 0..<numBuffers {
-            let bufferRef = UnsafeMutablePointer<AudioQueueBufferRef?>.allocate(capacity: 1)
-            AudioQueueAllocateBuffer(queue, bufferSize, bufferRef)
-            if let buffer = bufferRef.pointee {
-                AudioQueueEnqueueBuffer(queue, buffer, 0, nil)
+        shouldBeListening = true
+
+        let audioSession = AVAudioSession.sharedInstance()
+        // Only check if it's denied, permission will be automatically asked.
+        if audioSession.recordPermission() == .denied {
+            throw PorcupineManagerPermissionError.recordingDenied
+        }
+
+        guard !isListening else {
+            return
+        }
+
+        let inputNode = audioEngine.inputNode
+
+        try audioSession.setCategory(AVAudioSessionCategoryRecord)
+        try audioSession.setMode(AVAudioSessionModeMeasurement)
+        try audioSession.setActive(true, with: .notifyOthersOnDeactivation)
+
+        let frameLength = UInt32(pv_porcupine_frame_length())
+        let sampleRate = Double(pv_sample_rate())
+
+        let recordingFormat = inputNode.inputFormat(forBus: busNumber)
+        let numberOfBuffer: UInt32 = 4
```

Comment:
I think in watchOS binding I was reading 4 buffer at a time because of a technical issue with watchOS. Here is a brief explanation: Porcupine reads audio in 32 ms frames. watchOS for some reason does not work if you set the buffer size to anything below 100 ms. I can't recall why. So for watchOS, we had to put a hack to read  4 * 32 = 128 ms buffers. This is not ideal and if possible we should avoid it on iOS as it increases the delay. Let me know what you think.

Reply:
Yes, this work, the buffer size is set to approx 8196, it's bigger than previously assigned 5644 (0.128 * recordingFormat.sampleRate, which is 44100).

I have tested on both iOS and watchOS.

Reply:
Oops, I misread this. I think I got what you mean now. I have to fix up the buffer size then.